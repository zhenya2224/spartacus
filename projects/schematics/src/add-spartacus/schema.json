{
  "$schema": "http://json-schema.org/schema",
  "$id": "SpartacusSchematics",
  "title": "Spartacus Schematics",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "$default": {
        "$source": "projectName"
      }
    },
    "features": {
      "type": "array",
      "uniqueItems": true,
      "items": {
        "enum": [
          "ASM",
          "Import-Export",
          "Saved-Cart",
          "Quick-Order",
          "CDC",
          "CDS",
          "Checkout",
          "Order",
          "Digital-Payments",
          "Administration",
          "Order-Approval",
          "Bulk-Pricing",
          "Image-Zoom",
          "Product-Variants",
          "VC-Configurator",
          "Textfield-Configurator",
          "CPQ-Configurator",
          "Qualtrics",
          "SmartEdit",
          "Store-Finder",
          "Personalization",
          "TMS-GTM",
          "TMS-AEPL",
          "Account",
          "Profile"
        ],
        "type": "string"
      },
      "default": [
        "ASM",
        "Checkout",
        "Order",
        "Import-Export",
        "Saved-Cart",
        "Quick-Order",
        "Image-Zoom",
        "Product-Variants",
        "SmartEdit",
        "Store-Finder",
        "Personalization",
        "Account",
        "Profile"
      ],
      "x-prompt": {
        "message": "Which Spartacus features would you like to set up?\nPlease note that for most Spartacus features to be properly configured, the Account feature is required.",
        "type": "list",
        "items": [
          {
            "value": "ASM",
            "label": "Assisted Services Module"
          },
          {
            "value": "Import-Export",
            "label": "Import/Export"
          },
          {
            "value": "Saved-Cart",
            "label": "Saved Cart"
          },
          {
            "value": "Quick-Order",
            "label": "Quick Order"
          },
          {
            "value": "CDC",
            "label": "Customer Data Cloud Integration"
          },
          {
            "value": "CDS",
            "label": "Context-Driven Services Integration"
          },
          {
            "value": "Checkout",
            "label": "Checkout"
          },
          {
            "value": "Order",
            "label": "Order"
          },
          {
            "value": "Digital-Payments",
            "label": "Digital Payments Integration"
          },
          {
            "value": "Administration",
            "label": "Organization - Adminstration (b2b feature)"
          },
          {
            "value": "Order-Approval",
            "label": "Organization - Order Approval (b2b feature)"
          },
          {
            "value": "Bulk-Pricing",
            "label": "Product - Bulk Pricing (b2b feature)"
          },
          {
            "value": "Product-Variants",
            "label": "Product - Variants"
          },
          {
            "value": "Image-Zoom",
            "label": "Product - Image Zoom"
          },
          {
            "value": "VC-Configurator",
            "label": "Product Configurator - Variant Configurator"
          },
          {
            "value": "Textfield-Configurator",
            "label": "Product Configurator - Textfield Configurator"
          },
          {
            "value": "CPQ-Configurator",
            "label": "Product Configurator - CPQ Configurator (b2b feature)"
          },
          {
            "value": "Qualtrics",
            "label": "Qualtrics"
          },
          {
            "value": "SmartEdit",
            "label": "SmartEdit"
          },
          {
            "value": "Store-Finder",
            "label": "Store Finder"
          },
          {
            "value": "Personalization",
            "label": "Tracking - Personalization"
          },
          {
            "value": "TMS-GTM",
            "label": "Tracking - Tag Management System - Google Tag Manager"
          },
          {
            "value": "TMS-AEPL",
            "label": "Tracking - Tag Management System - Adobe Experience Platform Launch"
          },
          {
            "value": "Account",
            "label": "User - Account"
          },
          {
            "value": "Profile",
            "label": "User - Profile"
          }
        ]
      }
    },
    "lazy": {
      "type": "boolean",
      "description": "Lazy load the Spartacus features.",
      "default": true
    },
    "baseUrl": {
      "type": "string",
      "description": "The url of the OCC backend",
      "default": "https://localhost:9002"
    },
    "occPrefix": {
      "type": "string",
      "description": "The OCC API prefix. E.g.: /occ/v2/"
    },
    "baseSite": {
      "type": "string",
      "description": "A comma-separated list of base site(s) to use with Spartacus"
    },
    "currency": {
      "type": "string",
      "description": "A comma-separated list of currencies to use in Spartacus",
      "default": "USD"
    },
    "language": {
      "type": "string",
      "description": "A comma-separated list of languages to use in Spartacus",
      "default": "en"
    },
    "urlParameters": {
      "type": "string",
      "description": "A comma-separated list that represents the order of site-context parameters. E.g.: ['baseSite', 'language', 'currency']"
    },
    "useMetaTags": {
      "type": "boolean",
      "description": "Configure baseUrl via meta tag in index.html"
    },
    "featureLevel": {
      "type": "string",
      "description": "Configure feature level for application."
    },
    "overwriteAppComponent": {
      "type": "boolean",
      "description": "Overwrite the contents of app.component.html file with the <cx-storefront> tag.",
      "default": true
    },
    "pwa": {
      "type": "boolean",
      "description": "Add PWA support to the project.",
      "default": false
    },
    "ssr": {
      "type": "boolean",
      "description": "Add Server-side rendering support to the project.",
      "default": false
    },
    "theme": {
      "type": "string",
      "description": "Select a style theme to add. E.g: santorini"
    }
  },
  "required": []
}
