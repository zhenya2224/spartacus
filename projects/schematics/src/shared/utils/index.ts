export * from './config-utils';
export * from './file-utils';
export * from './lib-utils';
export * from './module-file-utils';
export * from './new-module-utils';
export * from './package-utils';
export * from './program';
export * from './project-tsconfig-paths';
export * from './workspace-utils';
