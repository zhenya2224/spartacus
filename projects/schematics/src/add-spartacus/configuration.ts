import { Rule, SchematicsException, Tree } from '@angular-devkit/schematics';
import { SourceFile } from 'ts-morph';
import {
  FEATURES_CONFIG,
  I18N_CONFIG,
  OCC_CONFIG,
  PROVIDE_CONFIG_FUNCTION,
  SITE_CONTEXT_CONFIG,
  SPARTACUS_ASSETS,
  SPARTACUS_CONFIGURATION_MODULE,
  SPARTACUS_CORE,
  SPARTACUS_STOREFRONTLIB,
} from '../shared/constants';
import { addModuleProvider } from '../shared/utils/new-module-utils';
import { getSpartacusCurrentFeatureLevel } from '../shared/utils/package-utils';
import { createProgram, saveAndFormat } from '../shared/utils/program';
import { getProjectTsConfigPaths } from '../shared/utils/project-tsconfig-paths';
import { parseCSV } from '../shared/utils/transform-utils';
import { Schema as SpartacusOptions } from './schema';

export function addSpartacusConfiguration(options: SpartacusOptions): Rule {
  return (tree: Tree): Tree => {
    const { buildPaths } = getProjectTsConfigPaths(tree, options.project);

    if (!buildPaths.length) {
      throw new SchematicsException(
        'Could not find any tsconfig file. Cannot configure SpartacusConfigurationModule.'
      );
    }

    const basePath = process.cwd();
    for (const tsconfigPath of buildPaths) {
      addConfiguration(tree, tsconfigPath, basePath, options);
    }
    return tree;
  };
}

function addConfiguration(
  tree: Tree,
  tsconfigPath: string,
  basePath: string,
  options: SpartacusOptions
): void {
  const { appSourceFiles } = createProgram(tree, basePath, tsconfigPath);

  for (const sourceFile of appSourceFiles) {
    if (
      sourceFile
        .getFilePath()
        .includes(`${SPARTACUS_CONFIGURATION_MODULE}.module.ts`)
    ) {
      addCommonConfiguration(sourceFile, options);

      saveAndFormat(sourceFile);

      break;
    }
  }
}

function addCommonConfiguration(
  sourceFile: SourceFile,
  options: SpartacusOptions
): void {
  addModuleProvider(sourceFile, {
    import: [
      {
        moduleSpecifier: SPARTACUS_CORE,
        namedImports: [PROVIDE_CONFIG_FUNCTION],
      },
      {
        moduleSpecifier: SPARTACUS_STOREFRONTLIB,
        namedImports: ['layoutConfig'],
      },
    ],
    content: `provideConfig(layoutConfig)`,
  });
  addModuleProvider(sourceFile, {
    import: [
      {
        moduleSpecifier: SPARTACUS_CORE,
        namedImports: [PROVIDE_CONFIG_FUNCTION],
      },
      {
        moduleSpecifier: SPARTACUS_STOREFRONTLIB,
        namedImports: ['mediaConfig'],
      },
    ],
    content: `provideConfig(mediaConfig)`,
  });
  addModuleProvider(sourceFile, {
    import: {
      moduleSpecifier: SPARTACUS_STOREFRONTLIB,
      namedImports: ['defaultCmsContentProviders'],
    },
    content: `...defaultCmsContentProviders`,
  });

  addStorefrontConfig(sourceFile, options);
}

function createSiteContextConfig(options: SpartacusOptions): string {
  const currency = parseCSV(options.currency, ['USD']).toUpperCase();
  const language = parseCSV(options.language, ['en']).toLowerCase();
  let contextConfig = `
      context: {
        currency: [${currency}],
        language: [${language}],`;

  if (options.baseSite) {
    const baseSites = parseCSV(options.baseSite);
    contextConfig += `\nbaseSite: [${baseSites}],`;
  }

  if (options.urlParameters) {
    const urlParameters = parseCSV(options.urlParameters);
    contextConfig += `\nurlParameters: [${urlParameters}]`;
  }

  contextConfig += `},`;

  return `provideConfig(<${SITE_CONTEXT_CONFIG}>{${contextConfig}})`;
}

/**
 * Creates and adds a spartacus config based on the provided `options`.
 * @param options
 */
function addStorefrontConfig(
  sourceFile: SourceFile,
  options: SpartacusOptions
): void {
  const backendConfig = createBackendConfiguration(options);
  addModuleProvider(sourceFile, {
    import: [
      {
        moduleSpecifier: SPARTACUS_CORE,
        namedImports: [PROVIDE_CONFIG_FUNCTION, OCC_CONFIG],
      },
    ],
    content: backendConfig,
  });

  const siteContextConfig = createSiteContextConfig(options);
  addModuleProvider(sourceFile, {
    import: [
      {
        moduleSpecifier: SPARTACUS_CORE,
        namedImports: [PROVIDE_CONFIG_FUNCTION, SITE_CONTEXT_CONFIG],
      },
    ],
    content: siteContextConfig,
  });

  const i18nConfig = createI18NConfiguration();
  addModuleProvider(sourceFile, {
    import: [
      {
        moduleSpecifier: SPARTACUS_CORE,
        namedImports: [PROVIDE_CONFIG_FUNCTION, I18N_CONFIG],
      },
      {
        moduleSpecifier: SPARTACUS_ASSETS,
        namedImports: ['translations', 'translationChunksConfig'],
      },
    ],
    content: i18nConfig,
  });

  const featureLevelConfig = createFeatureLevelConfiguration(options);
  addModuleProvider(sourceFile, {
    import: [
      {
        moduleSpecifier: SPARTACUS_CORE,
        namedImports: [PROVIDE_CONFIG_FUNCTION, FEATURES_CONFIG],
      },
    ],
    content: featureLevelConfig,
  });
}

function createBackendConfiguration(options: SpartacusOptions): string {
  const baseUrlPart = `\n          baseUrl: '${options.baseUrl}',`;
  const occPrefixPart = options.occPrefix
    ? `prefix: '${options.occPrefix}'`
    : '';
  return `provideConfig(<${OCC_CONFIG}>{
    backend: {
      occ: {${options.useMetaTags ? '' : baseUrlPart}${occPrefixPart}
      }
    },
  })`;
}

function createI18NConfiguration(): string {
  return `provideConfig(<${I18N_CONFIG}>{
  i18n: {
    resources: translations,
    chunks: translationChunksConfig,
    fallbackLang: 'en'
  },
})`;
}

function createFeatureLevelConfiguration(options: SpartacusOptions): string {
  const featureLevelConfig = `
  features: {
    level: '${options.featureLevel || getSpartacusCurrentFeatureLevel()}'
  }`;
  return `provideConfig(<${FEATURES_CONFIG}>{${featureLevelConfig}})`;
}
