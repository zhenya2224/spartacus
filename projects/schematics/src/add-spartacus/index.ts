import {
  chain,
  noop,
  Rule,
  SchematicContext,
  SchematicsException,
  Tree,
} from '@angular-devkit/schematics';
import { NodeDependency } from '@schematics/angular/utility/dependencies';
import { WorkspaceProject } from '@schematics/angular/utility/workspace-models';
import { ANGULAR_HTTP, SPARTACUS_STOREFRONTLIB } from '../shared/constants';
import { getIndexHtmlPath } from '../shared/utils/file-utils';
import { appendHtmlElementToHead } from '../shared/utils/html-utils';
import {
  addPackageJsonDependencies,
  addSchematicsTasks,
  createSpartacusFeatureOptionsForLibrary,
  LibraryOptions,
  prepareCliPackageAndSubFeature,
} from '../shared/utils/lib-utils';
import { addModuleImport } from '../shared/utils/new-module-utils';
import {
  getPrefixedSpartacusSchematicsVersion,
  getSpartacusCurrentFeatureLevel,
  mapPackageToNodeDependencies,
  prepare3rdPartyDependencies,
  prepareSpartacusDependencies,
  readPackageJson,
} from '../shared/utils/package-utils';
import { createProgram, saveAndFormat } from '../shared/utils/program';
import { getProjectTsConfigPaths } from '../shared/utils/project-tsconfig-paths';
import {
  getDefaultProjectNameFromWorkspace,
  getProjectFromWorkspace,
  getProjectTargets,
  getWorkspace,
  scaffoldStructure,
} from '../shared/utils/workspace-utils';
import { addSpartacusConfiguration } from './configuration';
import { Schema as SpartacusOptions } from './schema';
import { setupSpartacusModule } from './spartacus';
import { setupSpartacusFeaturesModule } from './spartacus-features';
import { setupStoreModules } from './store';

function installStyles(options: SpartacusOptions): Rule {
  return (tree: Tree, context: SchematicContext): void => {
    const project = getProjectFromWorkspace(tree, options);
    const rootStyles = getProjectTargets(project)?.build?.options?.styles?.[0];
    const styleFilePath =
      typeof rootStyles === 'object'
        ? ((rootStyles as any)?.input as string)
        : rootStyles;

    if (!styleFilePath) {
      context.logger.warn(
        `Could not find the default style file for this project.`
      );
      context.logger.warn(
        `Please consider manually setting up spartacus styles`
      );
      return;
    }

    if (styleFilePath.split('.').pop() !== 'scss') {
      context.logger.warn(
        `Could not find the default SCSS style file for this project. `
      );
      context.logger.warn(
        `Please make sure your project is configured with SCSS and consider manually setting up spartacus styles.`
      );
      return;
    }

    const buffer = tree.read(styleFilePath);

    if (!buffer) {
      context.logger.warn(
        `Could not read the default style file within the project ${styleFilePath}`
      );
      context.logger.warn(
        `Please consider manually importing spartacus styles.`
      );
      return;
    }

    const htmlContent = buffer.toString();
    let insertion =
      '\n' +
      `$styleVersion: ${
        options.featureLevel || getSpartacusCurrentFeatureLevel()
      };\n@import '~@spartacus/styles/index';\n`;

    if (options?.theme) {
      insertion += `\n@import '~@spartacus/styles/scss/theme/${options.theme}';\n`;
    }

    if (htmlContent.includes(insertion)) {
      return;
    }

    const recorder = tree.beginUpdate(styleFilePath);

    recorder.insertLeft(htmlContent.length, insertion);
    tree.commitUpdate(recorder);
  };
}

function updateMainComponent(
  project: WorkspaceProject,
  options: SpartacusOptions
): Rule {
  return (host: Tree, context: SchematicContext): Tree | void => {
    const filePath = project.sourceRoot + '/app/app.component.html';
    const buffer = host.read(filePath);

    if (!buffer) {
      context.logger.warn(`Could not read app.component.html file.`);
      return;
    }

    const htmlContent = buffer.toString();
    const insertion = `<cx-storefront></cx-storefront>\n`;

    if (htmlContent.includes(insertion)) {
      return;
    }

    const recorder = host.beginUpdate(filePath);

    if (options && options.overwriteAppComponent) {
      recorder.remove(0, htmlContent.length);
      recorder.insertLeft(0, insertion);
    } else {
      recorder.insertLeft(htmlContent.length, `\n${insertion}`);
    }

    host.commitUpdate(recorder);

    return host;
  };
}

function updateIndexFile(tree: Tree, options: SpartacusOptions): Rule {
  return (host: Tree): Tree => {
    const projectIndexHtmlPath = getIndexHtmlPath(tree);
    const baseUrl = options.baseUrl || 'OCC_BACKEND_BASE_URL_VALUE';

    const metaTags = [
      `<meta name="occ-backend-base-url" content="${baseUrl}" />`,
      `<meta name="media-backend-base-url" content="MEDIA_BACKEND_BASE_URL_VALUE" />`,
    ];

    metaTags.forEach((metaTag) => {
      appendHtmlElementToHead(host, projectIndexHtmlPath, metaTag);
    });

    return host;
  };
}

function increaseBudgets(): Rule {
  return (tree: Tree): Tree => {
    const { path, workspace: angularJson } = getWorkspace(tree);
    const projectName = getDefaultProjectNameFromWorkspace(tree);

    const project = angularJson.projects[projectName];
    const architect = project.architect;
    const build = architect?.build;
    const configurations = build?.configurations;
    const productionConfiguration = configurations?.production;
    const productionBudgets = (
      ((productionConfiguration as any).budgets ?? []) as {
        type: string;
        maximumError: string;
      }[]
    ).map((budget) => {
      if (budget.type === 'initial') {
        return {
          ...budget,
          maximumError: '2.5mb',
        };
      }
      return budget;
    });

    const updatedAngularJson = {
      ...angularJson,
      projects: {
        ...angularJson.projects,
        [projectName]: {
          ...project,
          architect: {
            ...architect,
            build: {
              ...build,
              configurations: {
                ...configurations,
                production: {
                  ...productionConfiguration,
                  budgets: productionBudgets,
                },
              },
            },
          },
        },
      },
    };

    tree.overwrite(path, JSON.stringify(updatedAngularJson, null, 2));
    return tree;
  };
}

function prepareDependencies(): NodeDependency[] {
  const spartacusDependencies = prepareSpartacusDependencies();
  return spartacusDependencies.concat(prepare3rdPartyDependencies());
}

function updateAppModule(project: string): Rule {
  return (tree: Tree): Tree => {
    const { buildPaths } = getProjectTsConfigPaths(tree, project);

    if (!buildPaths.length) {
      throw new SchematicsException(
        'Could not find any tsconfig file. Cannot configure AppModule.'
      );
    }

    const basePath = process.cwd();
    for (const tsconfigPath of buildPaths) {
      const { appSourceFiles } = createProgram(tree, basePath, tsconfigPath);

      for (const sourceFile of appSourceFiles) {
        if (sourceFile.getFilePath().includes(`app.module.ts`)) {
          addModuleImport(sourceFile, {
            order: 1,
            import: {
              moduleSpecifier: ANGULAR_HTTP,
              namedImports: ['HttpClientModule'],
            },
            content: 'HttpClientModule',
          });
          addModuleImport(sourceFile, {
            order: 2,
            import: {
              moduleSpecifier: SPARTACUS_STOREFRONTLIB,
              namedImports: ['AppRoutingModule'],
            },
            content: 'AppRoutingModule',
          });

          saveAndFormat(sourceFile);

          break;
        }
      }
    }
    return tree;
  };
}

function addSpartacusFeatures(options: SpartacusOptions): Rule {
  return (tree: Tree, context: SchematicContext) => {
    const cliFeatures = prepareCliPackageAndSubFeature(options.features ?? []);
    const libraryOptions: LibraryOptions = {
      project: options.project,
      lazy: options.lazy,
    };
    const featureOptions = createSpartacusFeatureOptionsForLibrary(
      libraryOptions,
      cliFeatures
    );
    addSchematicsTasks(featureOptions, context);

    const packageJson = readPackageJson(tree);
    const spartacusVersion = getPrefixedSpartacusSchematicsVersion();
    const dependencies = Object.keys(cliFeatures).map((feature) =>
      mapPackageToNodeDependencies(feature, spartacusVersion)
    );
    return addPackageJsonDependencies(dependencies, packageJson)(tree, context);
  };
}

export function addSpartacus(options: SpartacusOptions): Rule {
  return (tree: Tree, context: SchematicContext) => {
    const project = getProjectFromWorkspace(tree, options);

    return chain([
      addPackageJsonDependencies(prepareDependencies(), readPackageJson(tree)),

      setupStoreModules(options.project),

      scaffoldStructure(options),

      setupSpartacusModule(options.project),

      setupSpartacusFeaturesModule(options.project),

      addSpartacusConfiguration(options),

      updateAppModule(options.project),
      installStyles(options),
      updateMainComponent(project, options),
      options.useMetaTags ? updateIndexFile(tree, options) : noop(),
      increaseBudgets(),

      addSpartacusFeatures(options),
    ])(tree, context);
  };
}
