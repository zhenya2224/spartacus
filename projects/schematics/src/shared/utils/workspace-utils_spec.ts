import { SchematicsException, Tree } from '@angular-devkit/schematics';
import {
  SchematicTestRunner,
  UnitTestTree,
} from '@angular-devkit/schematics/testing';
import {
  Schema as ApplicationOptions,
  Style,
} from '@schematics/angular/application/schema';
import {
  ProjectType,
  WorkspaceProject,
} from '@schematics/angular/utility/workspace-models';
import { Schema as WorkspaceOptions } from '@schematics/angular/workspace/schema';
import * as path from 'path';
import { Schema as SpartacusOptions } from '../../add-spartacus/schema';
import { SPARTACUS_CORE } from '../../shared/constants';
import {
  buildDefaultPath,
  getAngularJsonFile,
  getDefaultProjectNameFromWorkspace,
  getProject,
  getProjectFromWorkspace,
  getProjectTargets,
  getSourceRoot,
  getWorkspace,
  isWorkspaceProject,
  isWorkspaceSchema,
  validateSpartacusInstallation,
} from './workspace-utils';

const collectionPath = path.join(__dirname, '../../collection.json');
const schematicRunner = new SchematicTestRunner('schematics', collectionPath);

describe('Workspace utils', () => {
  let appTree: UnitTestTree;
  const workspaceOptions: WorkspaceOptions = {
    name: 'workspace',
    version: '0.5.0',
  };
  const appOptions: ApplicationOptions = {
    name: 'schematics-test',
    inlineStyle: false,
    inlineTemplate: false,
    routing: false,
    style: Style.Scss,
    skipTests: false,
    projectRoot: '',
  };
  const defaultOptions: SpartacusOptions = {
    project: 'schematics-test',
    lazy: true,
    features: [],
  };

  beforeEach(async () => {
    appTree = await schematicRunner
      .runExternalSchematicAsync(
        '@schematics/angular',
        'workspace',
        workspaceOptions
      )
      .toPromise();
    appTree = await schematicRunner
      .runExternalSchematicAsync(
        '@schematics/angular',
        'application',
        appOptions,
        appTree
      )
      .toPromise();
    appTree = await schematicRunner
      .runSchematicAsync('add-spartacus', defaultOptions, appTree)
      .toPromise();
  });

  describe('getSourceRoot', () => {
    it('should return the source root of the default project', async () => {
      const sourceRoot = getSourceRoot(appTree, {});
      expect(sourceRoot).toEqual('src');
    });
  });

  describe('getWorkspace', () => {
    it('should return data about project', async () => {
      const workspaceInfo = getWorkspace(appTree);
      expect(workspaceInfo.path).toEqual('/angular.json');
      expect(workspaceInfo.workspace.defaultProject).toEqual(appOptions.name);
    });
  });

  describe('getAngularJsonFile', () => {
    it('should return workspace', async () => {
      const workspace = getAngularJsonFile(appTree);
      expect(workspace.defaultProject).toEqual(appOptions.name);
    });

    it('should throw an error if Angular not found', async () => {
      expect(() => getAngularJsonFile(appTree, [])).toThrowError(
        new SchematicsException(`Could not find Angular`)
      );
    });

    it('should throw an error if not found path', async () => {
      expect(() =>
        getAngularJsonFile({
          ...appTree,
          read: (_path) => null,
          exists: (_path) => true,
        } as Tree)
      ).toThrowError(new SchematicsException(`Could not find (/angular.json)`));
    });
  });

  describe('getProjectFromWorkspace', () => {
    it('should return workspace project object', async () => {
      const workspaceProjectObject = getProjectFromWorkspace(
        appTree,
        defaultOptions
      );
      expect(workspaceProjectObject.projectType).toEqual('application');
      expect(workspaceProjectObject.sourceRoot).toEqual('src');
    });

    it('should throw an error if project is not passed', async () => {
      expect(() =>
        getProjectFromWorkspace(appTree as Tree, {} as SpartacusOptions)
      ).toThrowError(new SchematicsException('Option "project" is required.'));
    });

    it('should throw an error if project is not defined in this workspace', async () => {
      expect(() =>
        getProjectFromWorkspace(appTree, {
          project: 'projectKey',
        } as SpartacusOptions)
      ).toThrowError(
        new SchematicsException(`Project is not defined in this workspace.`)
      );
    });
  });

  describe('getDefaultProjectNameFromWorkspace', () => {
    it('should return default project name from workspace', () => {
      const workspaceName = getDefaultProjectNameFromWorkspace(appTree);
      expect(workspaceName).toEqual('schematics-test');
    });
  });

  describe('getProjectTargets', () => {
    it('should return project targets', () => {
      const projectTargets = getProjectTargets(
        getProjectFromWorkspace(appTree, defaultOptions)
      );
      expect(projectTargets).toMatchSnapshot();
    });

    it('should throw an error if not found', () => {
      expect(() =>
        getProjectTargets({
          projectType: ProjectType.Application,
          root: 'root',
          prefix: 'prefix',
          sourceRoot: '',
        })
      ).toThrowError(new Error('Project target not found.'));
    });
  });

  describe('buildDefaultPath', () => {
    let project: WorkspaceProject;
    beforeEach(() => {
      project = {
        projectType: ProjectType.Application,
        root: 'foo',
        prefix: 'app',
        sourceRoot: '',
      };
    });

    it('should handle projectType of application', () => {
      const result = buildDefaultPath(project);
      expect(result).toEqual('/foo/src/app');
    });

    it('should handle projectType of library', () => {
      project = { ...project, projectType: ProjectType.Library };
      const result = buildDefaultPath(project);
      expect(result).toEqual('/foo/src/lib');
    });

    it('should handle sourceRoot', () => {
      project = { ...project, sourceRoot: 'foo/bar/custom' };
      const result = buildDefaultPath(project);
      expect(result).toEqual('/foo/bar/custom/app');
    });
  });

  describe('getProject', () => {
    it('should return project', () => {
      const project = getProject(appTree, 'schematics-test');
      expect(project).toMatchSnapshot();
    });

    it('should return undefined for unknown name', () => {
      const project = getProject(appTree, 'unknownName');
      expect(project).toBeUndefined();
    });
  });

  describe('isWorkspaceSchema', () => {
    it('should return true', () => {
      const project = isWorkspaceSchema({
        projectType: ProjectType.Application,
        root: 'root',
        prefix: 'prefix',
        projects: {},
      });
      expect(project).toEqual(true);
    });

    it('should return false if it does not', () => {
      const project = isWorkspaceSchema({
        projectType: ProjectType.Application,
        root: 'root',
        prefix: 'prefix',
      });
      expect(project).toEqual(false);
    });
  });

  describe('isWorkspaceProject', () => {
    it('should return true for a compatible type', () => {
      expect(
        isWorkspaceProject({
          projectType: ProjectType.Application,
        })
      ).toEqual(true);
      expect(
        isWorkspaceProject({
          projectType: ProjectType.Library,
        })
      ).toEqual(true);
    });

    it('should return false for missing type', () => {
      expect(
        isWorkspaceProject({
          otherProperty: 'otherProperty',
        })
      ).toEqual(false);
      expect(isWorkspaceProject(null)).toEqual(false);
    });
  });

  describe('validateSpartacusInstallation', () => {
    it('should throw an error if key is missing', () => {
      expect(() =>
        validateSpartacusInstallation({ dependencies: {} })
      ).toThrowError(
        new SchematicsException(
          `Spartacus is not detected. Please first install Spartacus by running: 'ng add @spartacus/schematics'.
    To see more options, please check our documentation.`
        )
      );
    });

    it('should not throw an error if key exists', () => {
      expect(() =>
        validateSpartacusInstallation({
          dependencies: { [SPARTACUS_CORE]: '/..' },
        })
      ).not.toThrowError();
    });
  });
});
