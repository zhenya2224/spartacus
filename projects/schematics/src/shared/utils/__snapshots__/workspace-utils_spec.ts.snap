// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Workspace utils getProject should return project 1`] = `
Object {
  "architect": Object {
    "build": Object {
      "builder": "@angular-devkit/build-angular:browser",
      "configurations": Object {
        "development": Object {
          "buildOptimizer": false,
          "extractLicenses": false,
          "namedChunks": true,
          "optimization": false,
          "sourceMap": true,
          "vendorChunk": true,
        },
        "production": Object {
          "budgets": Array [
            Object {
              "maximumError": "2.5mb",
              "maximumWarning": "500kb",
              "type": "initial",
            },
            Object {
              "maximumError": "4kb",
              "maximumWarning": "2kb",
              "type": "anyComponentStyle",
            },
          ],
          "fileReplacements": Array [
            Object {
              "replace": "src/environments/environment.ts",
              "with": "src/environments/environment.prod.ts",
            },
          ],
          "outputHashing": "all",
        },
      },
      "defaultConfiguration": "production",
      "options": Object {
        "assets": Array [
          "src/favicon.ico",
          "src/assets",
        ],
        "index": "src/index.html",
        "inlineStyleLanguage": "scss",
        "main": "src/main.ts",
        "outputPath": "dist/schematics-test",
        "polyfills": "src/polyfills.ts",
        "scripts": Array [],
        "styles": Array [
          "src/styles.scss",
        ],
        "tsConfig": "tsconfig.app.json",
      },
    },
    "extract-i18n": Object {
      "builder": "@angular-devkit/build-angular:extract-i18n",
      "options": Object {
        "browserTarget": "schematics-test:build",
      },
    },
    "serve": Object {
      "builder": "@angular-devkit/build-angular:dev-server",
      "configurations": Object {
        "development": Object {
          "browserTarget": "schematics-test:build:development",
        },
        "production": Object {
          "browserTarget": "schematics-test:build:production",
        },
      },
      "defaultConfiguration": "development",
    },
    "test": Object {
      "builder": "@angular-devkit/build-angular:karma",
      "options": Object {
        "assets": Array [
          "src/favicon.ico",
          "src/assets",
        ],
        "inlineStyleLanguage": "scss",
        "karmaConfig": "karma.conf.js",
        "main": "src/test.ts",
        "polyfills": "src/polyfills.ts",
        "scripts": Array [],
        "styles": Array [
          "src/styles.scss",
        ],
        "tsConfig": "tsconfig.spec.json",
      },
    },
  },
  "prefix": "app",
  "projectType": "application",
  "root": "",
  "schematics": Object {
    "@schematics/angular:application": Object {
      "strict": true,
    },
    "@schematics/angular:component": Object {
      "style": "scss",
    },
  },
  "sourceRoot": "src",
}
`;

exports[`Workspace utils getProjectTargets should return project targets 1`] = `
Object {
  "build": Object {
    "builder": "@angular-devkit/build-angular:browser",
    "configurations": Object {
      "development": Object {
        "buildOptimizer": false,
        "extractLicenses": false,
        "namedChunks": true,
        "optimization": false,
        "sourceMap": true,
        "vendorChunk": true,
      },
      "production": Object {
        "budgets": Array [
          Object {
            "maximumError": "2.5mb",
            "maximumWarning": "500kb",
            "type": "initial",
          },
          Object {
            "maximumError": "4kb",
            "maximumWarning": "2kb",
            "type": "anyComponentStyle",
          },
        ],
        "fileReplacements": Array [
          Object {
            "replace": "src/environments/environment.ts",
            "with": "src/environments/environment.prod.ts",
          },
        ],
        "outputHashing": "all",
      },
    },
    "defaultConfiguration": "production",
    "options": Object {
      "assets": Array [
        "src/favicon.ico",
        "src/assets",
      ],
      "index": "src/index.html",
      "inlineStyleLanguage": "scss",
      "main": "src/main.ts",
      "outputPath": "dist/schematics-test",
      "polyfills": "src/polyfills.ts",
      "scripts": Array [],
      "styles": Array [
        "src/styles.scss",
      ],
      "tsConfig": "tsconfig.app.json",
    },
  },
  "extract-i18n": Object {
    "builder": "@angular-devkit/build-angular:extract-i18n",
    "options": Object {
      "browserTarget": "schematics-test:build",
    },
  },
  "serve": Object {
    "builder": "@angular-devkit/build-angular:dev-server",
    "configurations": Object {
      "development": Object {
        "browserTarget": "schematics-test:build:development",
      },
      "production": Object {
        "browserTarget": "schematics-test:build:production",
      },
    },
    "defaultConfiguration": "development",
  },
  "test": Object {
    "builder": "@angular-devkit/build-angular:karma",
    "options": Object {
      "assets": Array [
        "src/favicon.ico",
        "src/assets",
      ],
      "inlineStyleLanguage": "scss",
      "karmaConfig": "karma.conf.js",
      "main": "src/test.ts",
      "polyfills": "src/polyfills.ts",
      "scripts": Array [],
      "styles": Array [
        "src/styles.scss",
      ],
      "tsConfig": "tsconfig.spec.json",
    },
  },
}
`;
